proposals/lip-0052.md

```diff
@@ -975,103 +1509,479 @@ transferCrossChain(senderAddress,
+
     nftID = canonicalNFTID(nftID)
-    if NFTStore(nftID).lockingModuleID != NFT_NOT_LOCKED:
-        lock fails
 
-    NFTStore(nftID).lockingModuleID = moduleID
+    if NFTStore(nftID).lockingModule != NFT_NOT_LOCKED:
+        emitFailedLockEvent(module, nftID, RESULT_NFT_ALREADY_LOCKED)
+        raise Exception("NFT is already locked.")
+
+    NFTStore(nftID).lockingModule = module
+
+    emitEvent(
+        module = MODULE_NAME_NFT,
+        name = EVENT_NAME_LOCK,
+        data = {
+            "module": module,
+            "nftID": nftID,
+            "result": RESULT_SUCCESSFUL
+        },
+        topics = []
+    )
+
+    def emitFailedLockEvent(
+        module: Module,
+        nftID: NFTID
+        result: uint32
+    ) -> None:
+
+    emitPersistentEvent(
+        module = MODULE_NAME_NFT,
+        name = EVENT_NAME_LOCK,
+        data = {
+            "module": module,
+            "nftID": nftID,
+            "result": result
+        },
+        topics = []
+    )
+
 ```
 
 #### unlock
 
-This function unlocks an NFT that was locked to a module ID.
+This function unlocks an NFT that was locked to a module.
 
 ```python
-unlock(moduleID, nftID):
+def unlock(
+    module: Module,
+    nftID: NFTID
+) -> None:
+
     nftID = canonicalNFTID(nftID)
-    if NFTStore(nftID).lockingModuleID != moduleID:
-        unlock fails
 
-    NFTStore(nftID).lockingModuleID = NFT_NOT_LOCKED
+    if NFTStore(nftID).lockingModule != module:
+        emitFailedUnlockEvent(module, nftID, RESULT_UNAUTHORIZED_UNLOCK)
+        raise Exception("Unlocking NFT via module that didn't lock it.")
+
+    NFTStore(nftID).lockingModule = NFT_NOT_LOCKED
+
+    emitEvent(
+        module = MODULE_NAME_NFT,
+        name = EVENT_NAME_UNLOCK,
+        data = {
+            "module": module,
+            "nftID": nftID,
+            "result": RESULT_SUCCESSFUL
+        },
+        topics = []
+    )
+
+    def emitFailedUnlockEvent(
+        module: Module,
+        nftID: NFTID
+        result: uint32
+    ) -> None:
+
+    emitPersistentEvent(
+        module = MODULE_NAME_NFT,
+        name = EVENT_NAME_UNLOCK,
+        data = {
+            "module": module,
+            "nftID": nftID,
+            "result": result
+        },
+        topics = []
+    )
 ```
 
 #### setAttributes
 
-This function modifies the attributes of NFTs.
+This function modifies the `attributes` of NFTs.
 
 ```python
-setAttributes(newAttributes, nftID):
-    nftID = canonicalNFTID(nftID)
-    if (NFTStore(nftID) does not exist
-        or length(newAttributes) > MAX_BYTE_LENGTH_ATTRIBUTES):
-        setAttributes fails
-
-    NFTStore(nftID).attributes = newAttributes
+def setAttributes(
+    attributes: bytes,
+    nftID: NFTID
+) -> None:
+
+    if NFTStore(nftID) does not exist:
+        emitFailedsetAttributesEvent(attributes, nftID, RESULT_NFT_DOES_NOT_EXIST)
+        raise Exception("NFT substore entry does not exist.")
+
+    if length(attributes) > MAX_BYTE_LENGTH_ATTRIBUTES:
```

Comment:
minor point: I would make this check before, so in some cases we do not even have to read from the state

Reply:
What do you mean by before?